{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a63e47f-8541-4552-8d84-bb1d8490807d",
   "metadata": {},
   "source": [
    "# Workflow for deepTools ChIP-Seq Heatmap Generation\n",
    "\n",
    "The following notebook shows the optimized workflow for creating heatmaps starting from wig.gz files. Will use a combination of UCSC and deepTools utilities."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33402582-be79-4c39-bfc5-1a981038d825",
   "metadata": {},
   "source": [
    "## Tools download and setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1517068c-99ab-49dd-9d4d-34aec0094c6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the python program wigToBigWig from the UCSC's utilties library.\n",
    "\n",
    "!wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/wigToBigWig\n",
    "!chmod +x ./wigToBigWig\n",
    "\n",
    "#If you haven't done so yet, download deepTools using the following line\n",
    "\n",
    "!pip install deeptools==3.5.4.post1\n",
    "\n",
    "#When generating BigWig files, will need a file containing chromosome sizes\n",
    "\n",
    "!wget https://hgdownload.cse.ucsc.edu/goldenpath/mm10/bigZips/mm10.chrom.sizes\n",
    "chrom_size = \"mm10.chrom.sizes\"\n",
    "\n",
    "#For sorting bed files for filtering, downlaod bedops\n",
    "\n",
    "!conda install bioconda::bedops"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "294e12fd-5b1a-412c-bd52-96b597348690",
   "metadata": {},
   "source": [
    "## Preparing Wig Files for deepTools Heatmaps Creation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f8fbe71-a4ce-4f8e-a52a-6e426274af53",
   "metadata": {},
   "source": [
    "### Wig Files List Generator\n",
    "\n",
    "To automate the process, we will generate a list that will find the right file according to a sample_key.txt file and implement it accordingly. In our dataset, the wig files were labeled using an index code that would only be deciphered using the sample_key to figure out histone modification mark and their experimental group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4c7e53c-6060-47c3-b99a-e57203635c24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>assays</th>\n",
       "      <th>species</th>\n",
       "      <th>mark</th>\n",
       "      <th>sample name</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>SSM2_09June23_D_H3K4me1</td>\n",
       "      <td>PX3242_TGACCA-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>SSM2_09June23_D_H3K4me3</td>\n",
       "      <td>PX3242_CTTGTA-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>SSM2_09June23_D_H3K9me3</td>\n",
       "      <td>PX3242_AAGCGA-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>SSM2_09June23_D_H3K27me3</td>\n",
       "      <td>PX3242_ACTCTC-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>SSM2_09June23_D_H3K36me3</td>\n",
       "      <td>PX3242_ATACGG-CAGTACAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     project   assays species      mark               sample name  \\\n",
       "0  Underhill  Chipseq   mouse   H3K4me1   SSM2_09June23_D_H3K4me1   \n",
       "1  Underhill  Chipseq   mouse   H3K4me3   SSM2_09June23_D_H3K4me3   \n",
       "2  Underhill  Chipseq   mouse   H3K9me3   SSM2_09June23_D_H3K9me3   \n",
       "3  Underhill  Chipseq   mouse  H3K27me3  SSM2_09June23_D_H3K27me3   \n",
       "4  Underhill  Chipseq   mouse  H3K36me3  SSM2_09June23_D_H3K36me3   \n",
       "\n",
       "                    index  \n",
       "0  PX3242_TGACCA-CAGTACAG  \n",
       "1  PX3242_CTTGTA-CAGTACAG  \n",
       "2  PX3242_AAGCGA-CAGTACAG  \n",
       "3  PX3242_ACTCTC-CAGTACAG  \n",
       "4  PX3242_ATACGG-CAGTACAG  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# sample_key.txt (file with name conversion and other info) has been changed to a csv file through excel\n",
    "\n",
    "key = pd.read_csv(\"combined_sample_key.csv\")\n",
    "key.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "25190754-1d53-430e-9e23-bbabf5008e36",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>assays</th>\n",
       "      <th>species</th>\n",
       "      <th>mark</th>\n",
       "      <th>sample name</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>SSM2_09June23_D_H3K4me1</td>\n",
       "      <td>PX3242_TGACCA-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>SSM2_09June23_D_H3K4me3</td>\n",
       "      <td>PX3242_CTTGTA-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>SSM2_09June23_D_H3K9me3</td>\n",
       "      <td>PX3242_AAGCGA-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>SSM2_09June23_D_H3K27me3</td>\n",
       "      <td>PX3242_ACTCTC-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>SSM2_09June23_D_H3K36me3</td>\n",
       "      <td>PX3242_ATACGG-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>SSM2_09June23_D_H3K27ac</td>\n",
       "      <td>PX3242_CACGAT-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K36me2</td>\n",
       "      <td>SSM2_09June23_D_H3K36me2</td>\n",
       "      <td>PX3242_GAGTGG-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>Ubiquityl</td>\n",
       "      <td>SSM2_09June23_D_Ubiquityl</td>\n",
       "      <td>PX3242_CTGCTG-CAGTACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>Input</td>\n",
       "      <td>SSM2_09June23_D_Input</td>\n",
       "      <td>PX3242_TTCTCC-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_H3K4me1</td>\n",
       "      <td>PX3242_ACAGTG-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_H3K4me3</td>\n",
       "      <td>PX3242_AAACAT-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_H3K9me3</td>\n",
       "      <td>PX3242_AAGGAC-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_H3K27me3</td>\n",
       "      <td>PX3242_ACTGAT-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_H3K36me3</td>\n",
       "      <td>PX3242_ATCCTA-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_H3K27ac</td>\n",
       "      <td>PX3242_CACTCA-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K36me2</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_H3K36me2</td>\n",
       "      <td>PX3242_CCGCAA-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>Ubiquityl</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_Ubiquityl</td>\n",
       "      <td>PX3242_GAAACC-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>Input</td>\n",
       "      <td>CTL_Tongue_10Aug23_G_Input</td>\n",
       "      <td>PX3242_GTGAAA-GTAGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.H3K4me1.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.H3K4me3.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.H3K9me3.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.H3K27me3.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.H3K36me3.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.H3K27ac.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>H3K36me2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.H3K36me2.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>Ubiquityl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.H2AK119ub.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Underhill</td>\n",
       "      <td>Chipseq</td>\n",
       "      <td>mouse</td>\n",
       "      <td>Input</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E12_5FL.Input.q5.F1028.PET</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      project   assays species       mark                     sample name  \\\n",
       "0   Underhill  Chipseq   mouse    H3K4me1         SSM2_09June23_D_H3K4me1   \n",
       "1   Underhill  Chipseq   mouse    H3K4me3         SSM2_09June23_D_H3K4me3   \n",
       "2   Underhill  Chipseq   mouse    H3K9me3         SSM2_09June23_D_H3K9me3   \n",
       "3   Underhill  Chipseq   mouse   H3K27me3        SSM2_09June23_D_H3K27me3   \n",
       "4   Underhill  Chipseq   mouse   H3K36me3        SSM2_09June23_D_H3K36me3   \n",
       "5   Underhill  Chipseq   mouse    H3K27ac         SSM2_09June23_D_H3K27ac   \n",
       "6   Underhill  Chipseq   mouse   H3K36me2        SSM2_09June23_D_H3K36me2   \n",
       "7   Underhill  Chipseq   mouse  Ubiquityl       SSM2_09June23_D_Ubiquityl   \n",
       "8   Underhill  Chipseq   mouse      Input           SSM2_09June23_D_Input   \n",
       "9   Underhill  Chipseq   mouse    H3K4me1    CTL_Tongue_10Aug23_G_H3K4me1   \n",
       "10  Underhill  Chipseq   mouse    H3K4me3    CTL_Tongue_10Aug23_G_H3K4me3   \n",
       "11  Underhill  Chipseq   mouse    H3K9me3    CTL_Tongue_10Aug23_G_H3K9me3   \n",
       "12  Underhill  Chipseq   mouse   H3K27me3   CTL_Tongue_10Aug23_G_H3K27me3   \n",
       "13  Underhill  Chipseq   mouse   H3K36me3   CTL_Tongue_10Aug23_G_H3K36me3   \n",
       "14  Underhill  Chipseq   mouse    H3K27ac    CTL_Tongue_10Aug23_G_H3K27ac   \n",
       "15  Underhill  Chipseq   mouse   H3K36me2   CTL_Tongue_10Aug23_G_H3K36me2   \n",
       "16  Underhill  Chipseq   mouse  Ubiquityl  CTL_Tongue_10Aug23_G_Ubiquityl   \n",
       "17  Underhill  Chipseq   mouse      Input      CTL_Tongue_10Aug23_G_Input   \n",
       "18  Underhill  Chipseq   mouse    H3K4me1                             NaN   \n",
       "19  Underhill  Chipseq   mouse    H3K4me3                             NaN   \n",
       "20  Underhill  Chipseq   mouse    H3K9me3                             NaN   \n",
       "21  Underhill  Chipseq   mouse   H3K27me3                             NaN   \n",
       "22  Underhill  Chipseq   mouse   H3K36me3                             NaN   \n",
       "23  Underhill  Chipseq   mouse    H3K27ac                             NaN   \n",
       "24  Underhill  Chipseq   mouse   H3K36me2                             NaN   \n",
       "25  Underhill  Chipseq   mouse  Ubiquityl                             NaN   \n",
       "26  Underhill  Chipseq   mouse      Input                             NaN   \n",
       "\n",
       "                             index  \n",
       "0           PX3242_TGACCA-CAGTACAG  \n",
       "1           PX3242_CTTGTA-CAGTACAG  \n",
       "2           PX3242_AAGCGA-CAGTACAG  \n",
       "3           PX3242_ACTCTC-CAGTACAG  \n",
       "4           PX3242_ATACGG-CAGTACAG  \n",
       "5           PX3242_CACGAT-CAGTACAG  \n",
       "6           PX3242_GAGTGG-CAGTACAG  \n",
       "7           PX3242_CTGCTG-CAGTACAG  \n",
       "8           PX3242_TTCTCC-GTAGTACT  \n",
       "9           PX3242_ACAGTG-GTAGTACT  \n",
       "10          PX3242_AAACAT-GTAGTACT  \n",
       "11          PX3242_AAGGAC-GTAGTACT  \n",
       "12          PX3242_ACTGAT-GTAGTACT  \n",
       "13          PX3242_ATCCTA-GTAGTACT  \n",
       "14          PX3242_CACTCA-GTAGTACT  \n",
       "15          PX3242_CCGCAA-GTAGTACT  \n",
       "16          PX3242_GAAACC-GTAGTACT  \n",
       "17          PX3242_GTGAAA-GTAGTACT  \n",
       "18    E12_5FL.H3K4me1.q5.F1028.PET  \n",
       "19    E12_5FL.H3K4me3.q5.F1028.PET  \n",
       "20    E12_5FL.H3K9me3.q5.F1028.PET  \n",
       "21   E12_5FL.H3K27me3.q5.F1028.PET  \n",
       "22   E12_5FL.H3K36me3.q5.F1028.PET  \n",
       "23    E12_5FL.H3K27ac.q5.F1028.PET  \n",
       "24   E12_5FL.H3K36me2.q5.F1028.PET  \n",
       "25  E12_5FL.H2AK119ub.q5.F1028.PET  \n",
       "26      E12_5FL.Input.q5.F1028.PET  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter the dataframe for only rows that correspond to ChIP-Seq data (marked as Chipseq in the assays column)\n",
    "\n",
    "ChIP_Seq = key[key[\"assays\"] == \"Chipseq\"]\n",
    "ChIP_Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e82cadb5-24bf-41ac-af63-bcaca909930c",
   "metadata": {},
   "source": [
    "### Converting Wig Files to BigWig\n",
    "\n",
    "BigWig (bw) files are necessary due to deepTools using that format as their input, but also they come in handy when utilzing IGV program. When trying to open wig files in IGV, they will reccomend converting it to TDF using their built-in tools, but that can take hours to process. Instead, converting to bw format takes only a couple of minutes. We will use the program wigToBigWig, however we must first unzip the files if that are compressed as .gz with a linux function called gzip."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "512c368a-81e0-4055-87fb-55af872b768e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using gzip to unzip the .gz compress wig files.\n",
    "# To automate this process, run a loop to gather all the names of the files from the samples_key file, and then feed those names through another loop that will unzip each one sequentially.\n",
    "\n",
    "compressed_list = []\n",
    "\n",
    "for i in range(len(ChIP_Seq[\"index\"])):\n",
    "    row = ChIP_Seq.iloc[i]\n",
    "    if \"E12_5\" in row[\"index\"]:\n",
    "        compressed_list.append(row[\"index\"] + \".wig.gz\")\n",
    "    else:\n",
    "        compressed_list.append(row[\"index\"] + \".markdup.q5.F3332.PET.wig.gz\")\n",
    "\n",
    "for name in compressed_list:\n",
    "    !gzip -dv -k {name}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9a51399-6371-448b-9d26-437ac4892f35",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Generating bw files from the wig files using UCSC's wigToBigWig tool\n",
    "\n",
    "chrom_size = \"mm10.chrom.sizes\"\n",
    "\n",
    "for name in ChIP_Seq['index']:\n",
    "    if \"E12_5\" in name:\n",
    "        !./wigToBigWig {name + \".wig\"} {chrom_size} {name + \".bw\"} -clip\n",
    "    else:\n",
    "        !./wigToBigWig {name + \".markdup.q5.F3332.PET.wig\"} {chrom_size} {name + \".markdup.q5.F3332.PET.bw\"} -clip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6898b71a-dadc-460f-b036-1b3ce2319664",
   "metadata": {},
   "source": [
    "### BigWig Files List Generator\n",
    "\n",
    "Create a list that will have the E12_5, CTL, and SSM2 bw files in a row for each corresponding histone type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "07d7e184-c00e-44b8-b8bc-e066d596d09c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'H3K27ac',\n",
       " 'H3K27me3',\n",
       " 'H3K36me2',\n",
       " 'H3K36me3',\n",
       " 'H3K4me1',\n",
       " 'H3K4me3',\n",
       " 'H3K9me3',\n",
       " 'Input',\n",
       " 'Ubiquityl'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find all the unique markers used in the ChIP-Seq data to form pairs of CTL vs SSM2 later (found in the mark column)\n",
    "\n",
    "marks = set(ChIP_Seq[\"mark\"])\n",
    "marks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d1549f96-ba06-4856-b550-e13ee39f6454",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make empty lists that will hold the string names of the markers, CTL index names, and the SSM2 index names\n",
    "\n",
    "mark_list = []\n",
    "E12_5_list = []\n",
    "CTL_list = []\n",
    "SSM2_list = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1aacda3a-b994-4f13-9c02-a237e008a749",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The following is a for loop that will ultimately scan, pull, and insert the index names of CTL and SSM2 bigwig files pairs\n",
    "\n",
    "for mark in marks:  # A for loop that will go through each unique marker name from the dataframe\n",
    "    E12_5 = ChIP_Seq[(ChIP_Seq[\"mark\"] == mark) & (ChIP_Seq[\"index\"].str.contains(\"E12_5\"))]\n",
    "    E12_5_index = E12_5[\"index\"].values[0]\n",
    "    CTL = ChIP_Seq[(ChIP_Seq[\"mark\"] == mark) & (ChIP_Seq[\"sample name\"].str.contains(\"CTL\"))] # Collect row that contains the right marker type and has the substring \"CTL\" in the sample name column\n",
    "    CTL_index = CTL[\"index\"].values[0] # Copy just the string of the index name from the CTL row. Have to use .values[0] or else you get a string that contains the row number and other useless data that will interfere with getting the right bigwig file name.\n",
    "    SSM2 = ChIP_Seq[(ChIP_Seq[\"mark\"] == mark) & (ChIP_Seq[\"sample name\"].str.contains(\"SSM2\"))] # Repeat of the CTL portion, but instead we find the row with the SSM2\n",
    "    SSM2_index = SSM2[\"index\"].values[0]\n",
    "    E12_5_bw = E12_5_index + \".bw\"\n",
    "    CTL_bw = CTL_index + \".markdup.q5.F3332.PET.bw\" # Add the remaining text from the bigwig file in order to call it properly\n",
    "    SSM2_bw = SSM2_index + \".markdup.q5.F3332.PET.bw\"\n",
    "    mark_list.append(mark) # Add the current marker name to the mark_list list\n",
    "    E12_5_list.append(E12_5_bw)\n",
    "    CTL_list.append(CTL_bw) # Add the full name of the CTL bigwig file to the CTL_list list\n",
    "    SSM2_list.append(SSM2_bw) # Add the full name of the SSM2 bigwig file to the SSM2_list list\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1f55d9d3-8bc3-4b2c-9cf9-3822b12a61a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mark</th>\n",
       "      <th>E12_5</th>\n",
       "      <th>CTL</th>\n",
       "      <th>SSM2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3K36me2</td>\n",
       "      <td>E12_5FL.H3K36me2.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_CCGCAA-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_GAGTGG-CAGTACAG.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>E12_5FL.H3K9me3.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_AAGGAC-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_AAGCGA-CAGTACAG.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Input</td>\n",
       "      <td>E12_5FL.Input.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_GTGAAA-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_TTCTCC-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ubiquityl</td>\n",
       "      <td>E12_5FL.H2AK119ub.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_GAAACC-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_CTGCTG-CAGTACAG.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>E12_5FL.H3K27ac.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_CACTCA-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_CACGAT-CAGTACAG.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>E12_5FL.H3K4me3.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_AAACAT-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_CTTGTA-CAGTACAG.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>E12_5FL.H3K27me3.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_ACTGAT-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_ACTCTC-CAGTACAG.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>E12_5FL.H3K36me3.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_ATCCTA-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_ATACGG-CAGTACAG.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>E12_5FL.H3K4me1.q5.F1028.PET.bw</td>\n",
       "      <td>PX3242_ACAGTG-GTAGTACT.markdup.q5.F3332.PET.bw</td>\n",
       "      <td>PX3242_TGACCA-CAGTACAG.markdup.q5.F3332.PET.bw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        mark                              E12_5  \\\n",
       "0   H3K36me2   E12_5FL.H3K36me2.q5.F1028.PET.bw   \n",
       "1    H3K9me3    E12_5FL.H3K9me3.q5.F1028.PET.bw   \n",
       "2      Input      E12_5FL.Input.q5.F1028.PET.bw   \n",
       "3  Ubiquityl  E12_5FL.H2AK119ub.q5.F1028.PET.bw   \n",
       "4    H3K27ac    E12_5FL.H3K27ac.q5.F1028.PET.bw   \n",
       "5    H3K4me3    E12_5FL.H3K4me3.q5.F1028.PET.bw   \n",
       "6   H3K27me3   E12_5FL.H3K27me3.q5.F1028.PET.bw   \n",
       "7   H3K36me3   E12_5FL.H3K36me3.q5.F1028.PET.bw   \n",
       "8    H3K4me1    E12_5FL.H3K4me1.q5.F1028.PET.bw   \n",
       "\n",
       "                                              CTL  \\\n",
       "0  PX3242_CCGCAA-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "1  PX3242_AAGGAC-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "2  PX3242_GTGAAA-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "3  PX3242_GAAACC-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "4  PX3242_CACTCA-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "5  PX3242_AAACAT-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "6  PX3242_ACTGAT-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "7  PX3242_ATCCTA-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "8  PX3242_ACAGTG-GTAGTACT.markdup.q5.F3332.PET.bw   \n",
       "\n",
       "                                             SSM2  \n",
       "0  PX3242_GAGTGG-CAGTACAG.markdup.q5.F3332.PET.bw  \n",
       "1  PX3242_AAGCGA-CAGTACAG.markdup.q5.F3332.PET.bw  \n",
       "2  PX3242_TTCTCC-GTAGTACT.markdup.q5.F3332.PET.bw  \n",
       "3  PX3242_CTGCTG-CAGTACAG.markdup.q5.F3332.PET.bw  \n",
       "4  PX3242_CACGAT-CAGTACAG.markdup.q5.F3332.PET.bw  \n",
       "5  PX3242_CTTGTA-CAGTACAG.markdup.q5.F3332.PET.bw  \n",
       "6  PX3242_ACTCTC-CAGTACAG.markdup.q5.F3332.PET.bw  \n",
       "7  PX3242_ATACGG-CAGTACAG.markdup.q5.F3332.PET.bw  \n",
       "8  PX3242_TGACCA-CAGTACAG.markdup.q5.F3332.PET.bw  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make a dataframe that will house columns for marker type, and CTL & SSM2 bigwig names.\n",
    "import pandas as pd\n",
    "\n",
    "bw_table = pd.DataFrame({'mark':mark_list, 'E12_5':E12_5_list, 'CTL':CTL_list, 'SSM2':SSM2_list})\n",
    "bw_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670794ac-d507-462d-9b2a-b07dd4b3d311",
   "metadata": {},
   "source": [
    "### Log2 normalizing BigWig files\n",
    "\n",
    "For some of the heatmaps, we will use log2 normalization, which requires using the deeptools tool bigwigCompare. Two different bw files are needed, treatment bw file and a input bw file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "898917e1-de66-40a2-a777-72e5278a40be",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#The following code is used if log2 normalization is needed for the heatmap.\n",
    "#Isolate the input row which will be used as the reference bw file during log2 normalization.\n",
    "\n",
    "input = bw_table[bw_table.eq(\"Input\").any(axis=1)]\n",
    "\n",
    "#Make a for loop that will run through the bw_table list and utilize deeptools' bigWigCompare to normalize every bw file to its input file (CTL or SSM2 respectively) with bin sizes of 25.\n",
    "#Takes ~40 mins to run log2 normalization from ~1 GB bw files. Normalized bw files are around 600 MB. Number of processors is set to 25 - change accordingly.\n",
    "\n",
    "for i in range(len(bw_table['mark'])):\n",
    "    if bw_table.loc[i, 'mark'] != \"Input\":\n",
    "        !bigwigCompare -b1 {bw_table.loc[i, 'E12_5']} -b2 {input['E12_5'].values[0]} -bs 25 --operation log2 -o {bw_table.loc[i,'mark'] + \"_E12_5_log2.bw\"} -p 25\n",
    "        !bigwigCompare -b1 {bw_table.loc[i, 'CTL']} -b2 {input['CTL'].values[0]} -bs 25 --operation log2 -o {bw_table.loc[i,'mark'] + \"_CTL_log2.bw\"} -p 25\n",
    "        !bigwigCompare -b1 {bw_table.loc[i, 'SSM2']} -b2 {input['SSM2'].values[0]} -bs 25 --operation log2 -o {bw_table.loc[i,'mark'] + \"_SSM2_log2.bw\"} -p 25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66bc54a9-504e-47c8-938e-aedec52a4fbf",
   "metadata": {},
   "source": [
    "## Generating Heatmaps\n",
    "\n",
    "Before creating a heatmap, we must first generate an intermediate matrix file that will house score values of different genes upstream and downstream of the region of interest based on the bw file and a reference region bed file (gene list)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf23dfab-622c-4ea5-b290-4862812bc1c5",
   "metadata": {},
   "source": [
    "We used a variety of bed file that either included all the genes found in the mm10 genomes or a curated gene list.\n",
    "1. For the former, it can be downloaded from [here](https://sourceforge.net/projects/rseqc/files/BED/Mouse_Mus_musculus/mm10_UCSC_knownGene.bed.gz/download).\n",
    "2. For the latter, it can be generated using [UCSC's table browser website](https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=2125137146_54BnotAR6qPQ2C8qFACtCzdewo7Y&clade=mammal&org=Mouse&db=mm10&hgta_group=genes&hgta_track=refSeqComposite&hgta_table=refGene&hgta_regionType=genome&position=chr12%3A56%2C694%2C976-56%2C714%2C605&hgta_outputType=primaryTable&hgta_outFileName=). From trial runs, the most polished bed file was generated using the NCBI refseq track and UCSC refseq table. For identifiers, input a txt file that contains a list of genes. Export as a bed file with no headers. Due to the gene data set, they will occassionally have multiple versions of the same gene list, creating repetitions of one single gene in the bed file. This will mess up the heatmap due to the pressence of large and broad streaks. To fix this, open the bed file using a txt file editor (for example: Notepad on Windows) and delete rows that have similar chromosome, start, and end values. To make things easier, should sort the bed file using bedops' sort_bed program."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c38bb41c-8ef1-4a23-95e0-d845819a1391",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import os\n",
    "\n",
    "#Prepare bed files containing the start and end positions of genes using the USCS table browser tool. Specifically, utilizing the NCBI dataset and UCSC refseq table.\n",
    "#Checked bed file using IGV.\n",
    "\n",
    "cluster_1 = \"mm10_cluster1_genelist_NCBI_UCSC.bed\"\n",
    "cluster_2 = \"mm10_cluster2_genelist_NCBI_UCSC.bed\"\n",
    "cluster_3 = \"mm10_cluster3_genelist_NCBI_UCSC.bed\"\n",
    "Hox_genes = \"mm10_HoxGenes_genelist_NCBI_UCSC.bed\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86c76502-02a3-4483-ab4a-4580b30ab0b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sorting gene list bed file to make filtering easier\n",
    "\n",
    "bed_list = [cluster_1, cluster_2, cluster_3, Hox_genes]\n",
    "\n",
    "for bed in bed_list:\n",
    "    !sort-bed {bed} > {\"sorted_\" + bed}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "017466ce-9fd4-42fd-a713-eb60f9187ad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removing duplicates of genes from the bed file based on exact same chromosome, start, and end.\n",
    "\n",
    "for bed in bed_list:\n",
    "    !sort -k1,1 -k2,2n -u {\"sorted_\" + bed} -o {\"filtered_sorted_\" + bed}\n",
    "    !sort -k1,1 -k3,3n -u {\"filtered_sorted_\" + bed} -o {\"filtered_sorted_\" + bed}\n",
    "\n",
    "# IMPORTANT\n",
    "# There will still be duplicates of genes that remain due to small differences in start and end values. Will have to go through it manually and remove them. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d3eef4d-3838-4d3d-97ae-cbba943a33d4",
   "metadata": {},
   "source": [
    "### Heatmaps around TSS, using absolute values, and all mm10 genes\n",
    "\n",
    "The following will generate pairs of heatmaps for each histone modification with the CTL group on the left and SSM2 group on the right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e3f527b-f2e4-4813-8010-486ba6076d2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using the bw_table, we will now go through each row/marker type and input the CTL and SSM2 bigwig files into the deeptools computeMatrix program\n",
    "# This will create a matrix from both CTL and SSM2 bigwigs that can later be used to create heatmaps\n",
    "\n",
    "for i in range(len(bw_table['mark'])):\n",
    "    !computeMatrix reference-point -S {bw_table.loc[i, 'CTL']} {bw_table.loc[i, 'SSM2']} -R mm10_UCSC_knownGene.bed -bs 25 -b 5000 -a 5000  -out {bw_table.loc[i, 'mark'] + \"_CTL_SSM2_referencePoint.matrix\"} --missingDataAsZero --skipZeros --samplesLabel {bw_table.loc[i, 'mark'] + \"_CTL\"} {bw_table.loc[i, 'mark'] + \"_SSM2\"} -p 20\n",
    "\n",
    "# Similar to the code above for running computeMatrix, we will use the bw_table to run deeptool's plotHeatmap\n",
    "# This will make heatmap pairs of CTL and SSM2\n",
    "\n",
    "for i in range(len(bw_table['mark'])):\n",
    "    !plotHeatmap -m {bw_table.loc[i, 'mark'] + \"_CTL_SSM2_referencePoint.matrix\"} --colorMap RdBu --heatmapHeight 25 --heatmapWidth 5 -out {bw_table.loc[i, 'mark'] + \"_CTL_SSM2_referencePoint.svg\"} --sortUsing max --regionsLabel genes --legendLocation none"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7dd9427-d215-4e3c-b6c1-446b4b0b62fc",
   "metadata": {},
   "source": [
    "### Heatmaps around TSS, using log2 normalized values, and all mm10 genes\n",
    "\n",
    "The following will generate pairs of heatmaps for each histone modification with the CTL group on the left and SSM2 group on the right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3320f0d-1eda-4956-bb32-e11d02f6f7db",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Using the bw_table, we will now go through each row/marker type and input the CTL and SSM2 bigwig files into the deeptools computeMatrix program\n",
    "# This will create a matrix from both CTL and SSM2 bigwigs that can later be used to create heatmaps\n",
    "\n",
    "for i in range(len(bw_table['mark'])):\n",
    "    !computeMatrix reference-point -S {bw_table.loc[i, 'mark'] + \"_CTL_log2.bw\"} {bw_table.loc[i, 'mark'] + \"_SSM2_log2.bw\"} -R mm10_UCSC_knownGene.bed -bs 25 -b 5000 -a 5000  -out {bw_table.loc[i, 'mark'] + \"_CTL_SSM2_log2_reference_point.matrix\"} --missingDataAsZero --skipZeros --samplesLabel {bw_table.loc[i, 'mark'] + \"_CTL\"} {bw_table.loc[i, 'mark'] + \"_SSM2\"} -p 15\n",
    "\n",
    "# Similar to the code above for running computeMatrix, we will use the bw_table to run deeptool's plotHeatmap\n",
    "# This will make heatmap pairs of CTL and SSM2\n",
    "\n",
    "for i in range(len(bw_table['mark'])):\n",
    "    !plotHeatmap -m {bw_table.loc[i, 'mark'] + \"_CTL_SSM2_log2_referencePoint.matrix\"} --colorMap RdBu --heatmapHeight 25 --heatmapWidth 5 -out {bw_table.loc[i, 'mark'] + \"_CTL_SSM2_log2_referencePoint.svg\"} --sortUsing max --regionsLabel genes --legendLocation none"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb87cd2-86d2-41bb-bf8f-f21d9db47ce1",
   "metadata": {},
   "source": [
    "### Combined heatmaps around TSS, using absolute values, and curated gene lists\n",
    "\n",
    "All the bw files and different histone modifications were implemented and processed together. 2 different heatmaps will ultimately be generated, with one utilizing 3 different bed files (cluster 1, cluster 2, and cluster 3), and the other only used a list of Hox genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daa54ff9-dfe2-4ead-9eab-12cfd18938a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#For the sake of easier name calling of bw files for combined heatmaps, we will rename them with more obvious names.\n",
    "#To do this, we will run a for loop on the bw_table and use the os.rename() function.\n",
    "\n",
    "for i in range(len(bw_table['mark'])):\n",
    "    for n in range(len(bw_table.columns)-1):\n",
    "        old_name = bw_table.loc[i, bw_table.columns[n+1]]\n",
    "        new_name = bw_table.loc[i, bw_table.columns[0]] + \"_\" + bw_table.columns[n+1] + \".bw\"\n",
    "        print(\"Old name = \" + old_name + \" New name = \" +  new_name)\n",
    "        os.rename(old_name, new_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fc9a79e-11e6-454a-9dc4-92e6d72316e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Rather than writing out 24 different file names and 24 different labels, we will automate the process.\n",
    "\n",
    "histone_order = ['H3K4me3', 'H3K27ac', 'H3K27me3', 'Ubiquityl', 'H3K4me1', 'H3K9me3', 'H3K36me2', 'H3K36me3'] #Specify the order of histone modifications.\n",
    "bw_list = ''\n",
    "histone_list = ''\n",
    "for histone in histone_order:\n",
    "    bw_list += (histone + \"_E12_5.bw \") #Specifiy the order of bw files per histone modification type (for this example: E12.5, CTL, then SSM2).\n",
    "    bw_list += (histone + \"_CTL.bw \")\n",
    "    bw_list += (histone + \"_SSM2.bw \")\n",
    "    histone_list += (histone + \"_E12_5 \") #The order of heatmap labels corresponding to the order of bw files.\n",
    "    histone_list += (histone + \"_CTL \")\n",
    "    histone_list += (histone + \"_SSM2 \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3657ebe8-19c4-4466-a972-245dcf743a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute a heatmap matrix that combines all 16 different bw files in a specific order and based around select regions using bed files.\n",
    "#The following will utilize non-normalized bw files.\n",
    "#Two separate heatmaps will be ultimately generated, one with 3 different clusters using gene lists, and one that uses Hox genes.\n",
    "#Will generate matrices that calculate scores 1 kb upstream and downstream of the transciption start site (TSS).\n",
    "\n",
    "!computeMatrix reference-point \\\n",
    "    -S {bw_list} \\\n",
    "    -R filtered_sorted_mm10_cluster1_genelist_NCBI_UCSC.bed filtered_sorted_mm10_cluster2_genelist_NCBI_UCSC.bed filtered_sorted_mm10_cluster3_genelist_NCBI_UCSC.bed \\\n",
    "    -bs 25 -b 5000 -a 5000  -out E12_5CTL_SSM2_TSS_5kb_clusters.matrix --missingDataAsZero --skipZeros \\\n",
    "    --samplesLabel {histone_list} \\\n",
    "    -p 25 --verbose\n",
    "\n",
    "!computeMatrix reference-point \\\n",
    "    -S {bw_list} \\\n",
    "    -R filtered_sorted_mm10_HoxGenes_genelist_NCBI_UCSC.bed \\\n",
    "    -bs 25 -b 5000 -a 5000  -out E12_5_CTL_SSM2_TSS_5kb_HoxGenes.matrix --missingDataAsZero --skipZeros \\\n",
    "    --samplesLabel {histone_list} \\\n",
    "    -p 25 --verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14afab08-841e-491c-9fb7-79b6ea49178c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate heatmaps in svg format for both types.\n",
    "\n",
    "!plotHeatmap -m E12_5_CTL_SSM2_TSS_5kb_clusters.matrix --colorMap RdBu --heatmapHeight 25 --heatmapWidth 5 \\\n",
    "    -out E12_5_CTL_SSM2_TSS_5kb_clusters.svg --sortUsing max --legendLocation none \\\n",
    "    --regionsLabel \"Cluster 1\" \"Cluster 2\" \"Cluster 3\" --whatToShow \"heatmap and colorbar\" --outFileSortedRegions \"E12_5_CTL_SSM2_TSS_5kb_clusters.bed\"\n",
    "\n",
    "plotHeatmap -m E12_5_CTL_SSM2_TSS_5kb_HoxGenes.matrix --colorMap RdBu --heatmapHeight 25 --heatmapWidth 5 \\\n",
    "    -out E12_5_CTL_SSM2_TSS_5kb_HoxGenes.svg --sortUsing max --legendLocation none \\\n",
    "    --regionsLabel \"Hox Genes\" --whatToShow \"heatmap and colorbar\" --outFileSortedRegions \"E12_5_CTL_SSM2_TSS_5kb_HoxGenes.bed\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12104e85-e576-4c90-af69-fa93ee450626",
   "metadata": {},
   "source": [
    "### Combined heatmaps around TSS, using log2 normailzed values, and curated gene lists\n",
    "\n",
    "All the bw files and different histone modifications were implemented and processed together. 2 different heatmaps will ultimately be generated, with one utilizing 3 different bed files (cluster 1, cluster 2, and cluster 3), and the other only used a list of Hox genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce79e6e6-2d16-4363-9574-141ffc306a4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Rather than writing out 24 different file names and 24 different labels, we will automate the process.\n",
    "\n",
    "histone_order = ['H3K4me3', 'H3K27ac', 'H3K27me3', 'Ubiquityl', 'H3K4me1', 'H3K9me3', 'H3K36me2', 'H3K36me3'] #Specify the order of histone modifications.\n",
    "bw_log2_list = ''\n",
    "histone_list = ''\n",
    "for histone in histone_order:\n",
    "    bw_log2_list += (histone + \"_E12_5_log2.bw \") #Specifiy the order of bw files per histone modification type (for this example: E12.5, CTL, then SSM2).\n",
    "    bw_log2_list += (histone + \"_CTL_log2.bw \")\n",
    "    bw_log2_list += (histone + \"_SSM2_log2.bw \")\n",
    "    histone_list += (histone + \"_E12_5 \") #The order of heatmap labels corresponding to the order of bw files.\n",
    "    histone_list += (histone + \"_CTL \")\n",
    "    histone_list += (histone + \"_SSM2 \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9264a0a1-d659-4294-b136-08e9a677b669",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Compute a heatmap matrix that combines all 16 different bw files in a specific order and based around select regions using bed files.\n",
    "#The following will utilize non-normalized bw files.\n",
    "#Two separate heatmaps will be ultimately generated, one with 3 different clusters using gene lists, and one that uses Hox genes.\n",
    "#Will generate matrices that calculate scores 1 kb upstream and downstream of the transciption start site (TSS).\n",
    "\n",
    "!computeMatrix reference-point \\\n",
    "    -S {bw_log2_list} \\\n",
    "    -R mm10_cluster1_genelist_NCBI_UCSC_filtered.bed mm10_cluster2_genelist_NCBI_UCSC_filtered.bed mm10_cluster3_genelist_NCBI_UCSC_filtered.bed -bs 25 -b 5000 -a 5000  \\\n",
    "    -out CTL_E12_5_SSM2_TSS_log2_5kb_clusters_filtered.matrix --missingDataAsZero --skipZeros \\\n",
    "    --samplesLabel {histone_list} \\\n",
    "    -p 25 --verbose\n",
    "\n",
    "!computeMatrix reference-point \\\n",
    "    -S {bw_log2_list} \\\n",
    "    -R mm10_HoxGenes_genelist_NCBI_UCSC_filtered.bed -bs 25 -b 5000 -a 5000  \\\n",
    "    -out CTL_E12_5_SSM2_TSS_log2_5kb_HoxGenes_filtered.matrix --missingDataAsZero --skipZeros \\\n",
    "    --samplesLabel {histone_list} \\\n",
    "    -p 25 --verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45e6761f-0011-4e1d-873c-8b6da75b9824",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate heatmaps in svg format for both types.\n",
    "\n",
    "!plotHeatmap -m E12_5_CTL_SSM2_TSS_log2_5kb_clusters_filtered.matrix --colorMap RdBu --heatmapHeight 25 --heatmapWidth 5 \\\n",
    "    -out E12_5_CTL_SSM2_TSS_log2_5kb_clusters_filtered.svg --sortUsing max --legendLocation none \\\n",
    "    --regionsLabel \"Cluster 1\" \"Cluster 2\" \"Cluster 3\" --whatToShow \"heatmap and colorbar\" --outFileSortedRegions \"E12_5_CTL_SSM2_TSS_5kb_log2_clusters.bed\"\n",
    "\n",
    "!plotHeatmap -m E12_5_CTL_SSM2_TSS_log2_5kb_HoxGenes_filtered.matrix --colorMap RdBu --heatmapHeight 10 --heatmapWidth 5 \\\n",
    "    -out E12_5_CTL_SSM2_TSS_log2_5kb_HoxGenes_filtered.svg --sortUsing max --legendLocation none \\\n",
    "    --regionsLabel \"Hox Genes\" --whatToShow \"heatmap and colorbar\" --outFileSortedRegions \"E12_5_CTL_SSM2_TSS_5kb_log2_HoxGenes.bed\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "658e0c20-b71a-4cf1-9d69-4c6ba7b70191",
   "metadata": {},
   "source": [
    "# Program Versions Used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c4ecadce-6507-4162-b68b-96dd329d3786",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accessible-pygments==0.0.4\n",
      "adjustText==0.8\n",
      "aiohttp==3.9.3\n",
      "aiohttp-cors==0.7.0\n",
      "aiosignal==1.3.1\n",
      "alabaster==0.7.13\n",
      "anndata==0.9.2\n",
      "annotated-types==0.6.0\n",
      "annoy==1.17.3\n",
      "anyio==4.2.0\n",
      "appdirs==1.4.4\n",
      "arboreto==0.1.6\n",
      "argon2-cffi==23.1.0\n",
      "argon2-cffi-bindings==21.2.0\n",
      "arrow==1.3.0\n",
      "asttokens @ file:///opt/conda/conda-bld/asttokens_1646925590279/work\n",
      "async-lru==2.0.4\n",
      "async-timeout==4.0.3\n",
      "attr==0.3.2\n",
      "attrs==23.2.0\n",
      "Babel==2.14.0\n",
      "backcall @ file:///home/ktietz/src/ci/backcall_1611930011877/work\n",
      "backports.zoneinfo==0.2.1\n",
      "bbknn==1.6.0\n",
      "beautifulsoup4==4.12.3\n",
      "bidict==0.22.1\n",
      "bioservices==1.11.2\n",
      "bleach==6.1.0\n",
      "blessed==1.20.0\n",
      "blosc2==2.0.0\n",
      "bokeh==3.1.1\n",
      "boltons==23.1.1\n",
      "bs4==0.0.2\n",
      "cachetools==5.3.2\n",
      "cattrs==23.2.3\n",
      "certifi==2024.2.2\n",
      "cffi==1.16.0\n",
      "charset-normalizer==3.3.2\n",
      "click==8.1.7\n",
      "cloudpickle==3.0.0\n",
      "colorama==0.4.6\n",
      "colorful==0.5.6\n",
      "colorlog==6.8.2\n",
      "comm==0.2.1\n",
      "contourpy==1.1.1\n",
      "cryptography==42.0.2\n",
      "ctxcore==0.2.0\n",
      "cycler==0.12.1\n",
      "Cython==0.29.37\n",
      "cytoolz==0.12.3\n",
      "dask==2023.5.0\n",
      "dataclasses-json==0.6.4\n",
      "debugpy @ file:///croot/debugpy_1690905042057/work\n",
      "decorator @ file:///opt/conda/conda-bld/decorator_1643638310831/work\n",
      "deepTools==3.5.4.post1\n",
      "deeptoolsintervals==0.1.9\n",
      "defusedxml==0.7.1\n",
      "dill==0.3.8\n",
      "distlib==0.3.8\n",
      "distributed==2023.5.0\n",
      "docutils==0.19\n",
      "easydev==0.12.1\n",
      "entrypoints==0.4\n",
      "et-xmlfile==1.1.0\n",
      "exceptiongroup @ file:///croot/exceptiongroup_1706031385326/work\n",
      "executing @ file:///opt/conda/conda-bld/executing_1646925071911/work\n",
      "fastjsonschema==2.19.1\n",
      "fbpca==1.0\n",
      "filelock==3.13.1\n",
      "fonttools==4.47.2\n",
      "fqdn==1.5.1\n",
      "frozendict==2.4.0\n",
      "frozenlist==1.4.1\n",
      "fsspec==2024.2.0\n",
      "future==0.18.3\n",
      "gensim==4.3.2\n",
      "geosketch==1.2\n",
      "get-annotations==0.1.2\n",
      "gevent==23.9.1\n",
      "globre==0.1.5\n",
      "google-api-core==2.16.2\n",
      "google-auth==2.27.0\n",
      "googleapis-common-protos==1.62.0\n",
      "gpustat==1.1.1\n",
      "greenlet==3.0.3\n",
      "grequests==0.7.0\n",
      "grpcio==1.60.1\n",
      "gseapy==0.10.8\n",
      "h11==0.14.0\n",
      "h5py==3.10.0\n",
      "harmonypy==0.0.9\n",
      "httpcore==1.0.2\n",
      "httpx==0.26.0\n",
      "idna==3.6\n",
      "igraph==0.10.8\n",
      "imageio==2.33.1\n",
      "imagesize==1.4.1\n",
      "importlib-metadata @ file:///croot/importlib_metadata-suite_1704813515092/work\n",
      "importlib-resources==6.1.1\n",
      "interlap==0.2.7\n",
      "intervaltree==3.1.0\n",
      "ipykernel @ file:///croot/ipykernel_1705933831282/work\n",
      "ipympl==0.9.3\n",
      "ipython @ file:///croot/ipython_1691532092695/work\n",
      "ipython-genutils==0.2.0\n",
      "ipywidgets==8.1.1\n",
      "isoduration==20.11.0\n",
      "jedi @ file:///tmp/build/80754af9/jedi_1644315233700/work\n",
      "Jinja2==3.1.3\n",
      "joblib==1.3.2\n",
      "json5==0.9.14\n",
      "jsonpickle==3.0.2\n",
      "jsonpointer==2.4\n",
      "jsonschema==4.21.1\n",
      "jsonschema-specifications==2023.12.1\n",
      "jupyter==1.0.0\n",
      "jupyter-console==6.6.3\n",
      "jupyter-events==0.9.0\n",
      "jupyter-lsp==2.2.2\n",
      "jupyter_client @ file:///croot/jupyter_client_1699455897726/work\n",
      "jupyter_core @ file:///croot/jupyter_core_1698937308754/work\n",
      "jupyter_server==2.12.5\n",
      "jupyter_server_terminals==0.5.2\n",
      "jupyterlab==4.1.0\n",
      "jupyterlab-widgets==3.0.9\n",
      "jupyterlab_pygments==0.3.0\n",
      "jupyterlab_server==2.25.2\n",
      "kaleido==0.2.1\n",
      "kiwisolver==1.4.5\n",
      "lazy_loader==0.3\n",
      "lckr_jupyterlab_variableinspector==3.2.1\n",
      "lda==2.0.0\n",
      "leidenalg @ file:///home/conda/feedstock_root/build_artifacts/leidenalg_1705928794442/work\n",
      "Levenshtein==0.24.0\n",
      "llvmlite==0.41.1\n",
      "locket==1.0.0\n",
      "loompy==3.0.7\n",
      "loomxpy @ git+https://github.com/aertslab/LoomXpy@61995ff10940968eac2cee8fe48300ab477a15d0\n",
      "louvain==0.8.1\n",
      "lxml==5.1.0\n",
      "lz4==4.3.3\n",
      "MACS2==2.2.9.1\n",
      "markdown-it-py==3.0.0\n",
      "MarkupSafe==2.1.5\n",
      "marshmallow==3.20.2\n",
      "matplotlib==3.7.4\n",
      "matplotlib-inline @ file:///opt/conda/conda-bld/matplotlib-inline_1662014470464/work\n",
      "mdurl==0.1.2\n",
      "mistune==3.0.2\n",
      "mizani==0.9.3\n",
      "msgpack==1.0.7\n",
      "mudata==0.2.3\n",
      "multidict==6.0.5\n",
      "multiprocessing-on-dill==3.5.0a4\n",
      "mypy-extensions==1.0.0\n",
      "natsort==8.4.0\n",
      "nbclient==0.9.0\n",
      "nbconvert==7.14.2\n",
      "nbformat==5.9.2\n",
      "nbsphinx==0.9.3\n",
      "nbsphinx-link==1.3.0\n",
      "ncls==0.0.68\n",
      "nest-asyncio @ file:///croot/nest-asyncio_1672387112409/work\n",
      "networkx==3.1\n",
      "notebook==7.0.7\n",
      "notebook_shim==0.2.3\n",
      "numba==0.58.1\n",
      "numexpr==2.8.6\n",
      "numpy==1.24.4\n",
      "numpy-groupies==0.9.22\n",
      "numpydoc==1.6.0\n",
      "nvidia-ml-py==12.535.133\n",
      "opencensus==0.11.4\n",
      "opencensus-context==0.1.3\n",
      "openpyxl==3.1.2\n",
      "overrides==7.7.0\n",
      "packaging @ file:///croot/packaging_1693575174725/work\n",
      "pandas==1.5.0\n",
      "pandoc==2.3\n",
      "pandocfilters==1.5.1\n",
      "papermill==2.5.0\n",
      "parso @ file:///opt/conda/conda-bld/parso_1641458642106/work\n",
      "partd==1.4.1\n",
      "patsy==0.5.6\n",
      "pbr==3.1.1\n",
      "pexpect @ file:///tmp/build/80754af9/pexpect_1605563209008/work\n",
      "pickleshare @ file:///tmp/build/80754af9/pickleshare_1606932040724/work\n",
      "pillow==10.2.0\n",
      "pkgutil_resolve_name==1.3.10\n",
      "platformdirs @ file:///croot/platformdirs_1692205439124/work\n",
      "plotly==5.18.0\n",
      "plotnine==0.12.4\n",
      "plumbum==1.8.2\n",
      "ply==3.11\n",
      "polars==0.20.7\n",
      "progressbar2==4.3.2\n",
      "prometheus-client==0.19.0\n",
      "prompt-toolkit @ file:///croot/prompt-toolkit_1704404351921/work\n",
      "protobuf==4.25.2\n",
      "psutil @ file:///opt/conda/conda-bld/psutil_1656431268089/work\n",
      "ptyprocess @ file:///tmp/build/80754af9/ptyprocess_1609355006118/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl\n",
      "pure-eval @ file:///opt/conda/conda-bld/pure_eval_1646925070566/work\n",
      "py-cpuinfo==9.0.0\n",
      "py-spy==0.3.14\n",
      "py2bit==0.3.0\n",
      "pyarrow==15.0.0\n",
      "pyasn1==0.5.1\n",
      "pyasn1-modules==0.3.0\n",
      "pybedtools==0.9.1\n",
      "pybigtools==0.1.0\n",
      "pyBigWig==0.3.22\n",
      "pybiomart==0.2.0\n",
      "pycistarget @ git+https://github.com/aertslab/pycistarget@81eb8757ec032a25db53b1952b8529064470e19f\n",
      "-e git+https://github.com/aertslab/pycisTopic.git@197fe045e78425dadeb589d39feea82ba577c15c#egg=pycisTopic\n",
      "pycparser==2.21\n",
      "pydantic==1.10.13\n",
      "pydantic_core==2.0.1\n",
      "pydata-sphinx-theme==0.14.4\n",
      "pyfasta==0.5.2\n",
      "pygam==0.9.0\n",
      "Pygments @ file:///croot/pygments_1684279966437/work\n",
      "pynndescent==0.5.11\n",
      "pyOpenSSL==24.0.0\n",
      "pyparsing==3.1.1\n",
      "pyranges==0.0.127\n",
      "pyrle==0.0.39\n",
      "pysam==0.22.0\n",
      "pyscenic @ git+https://github.com/aertslab/pySCENIC@1cd059f12c499b968dacd206b1dfaa1412709b3e\n",
      "python-dateutil @ file:///tmp/build/80754af9/python-dateutil_1626374649649/work\n",
      "python-igraph==0.11.3\n",
      "python-json-logger==2.0.7\n",
      "python-Levenshtein==0.24.0\n",
      "python-utils==3.8.2\n",
      "pytz==2024.1\n",
      "pyvis==0.3.2\n",
      "PyWavelets==1.4.1\n",
      "PyYAML==6.0.1\n",
      "pyzmq @ file:///croot/pyzmq_1705605076900/work\n",
      "qtconsole==5.5.1\n",
      "QtPy==2.4.1\n",
      "rapidfuzz==3.6.1\n",
      "ray==2.9.2\n",
      "referencing==0.33.0\n",
      "requests==2.31.0\n",
      "requests-cache==1.1.1\n",
      "rfc3339-validator==0.1.4\n",
      "rfc3986-validator==0.1.1\n",
      "rich==13.7.0\n",
      "rich-argparse==1.4.0\n",
      "rpds-py==0.17.1\n",
      "rsa==4.9\n",
      "scanorama==1.7.4\n",
      "scanpy==1.9.8\n",
      "scatac_fragment_tools @ git+https://github.com/aertslab/scatac_fragment_tools@a54d31f6b37a3f94b1507296c894bd9b4b4b79eb\n",
      "-e git+https://github.com/aertslab/scenicplus@e5ba6fcf42459b6e6b70e27359ddd11289d70cc5#egg=scenicplus\n",
      "scikit-image==0.21.0\n",
      "scikit-learn==1.3.2\n",
      "scipy==1.10.1\n",
      "scrublet==0.2.3\n",
      "seaborn==0.13.2\n",
      "Send2Trash==1.8.2\n",
      "session-info==1.0.0\n",
      "six @ file:///tmp/build/80754af9/six_1644875935023/work\n",
      "smart-open==6.4.0\n",
      "sniffio==1.3.0\n",
      "snowballstemmer==2.2.0\n",
      "sorted-nearest==0.0.39\n",
      "sortedcontainers==2.4.0\n",
      "soupsieve==2.5\n",
      "Sphinx==6.2.1\n",
      "sphinx-book-theme==1.0.1\n",
      "sphinx-rtd-theme==2.0.0\n",
      "sphinxcontrib-applehelp==1.0.4\n",
      "sphinxcontrib-devhelp==1.0.2\n",
      "sphinxcontrib-htmlhelp==2.0.1\n",
      "sphinxcontrib-jquery==4.1\n",
      "sphinxcontrib-jsmath==1.0.1\n",
      "sphinxcontrib-qthelp==1.0.3\n",
      "sphinxcontrib-serializinghtml==1.1.5\n",
      "stack-data @ file:///opt/conda/conda-bld/stack_data_1646927590127/work\n",
      "statistics==1.0.3.5\n",
      "statsmodels==0.14.1\n",
      "stdlib-list==0.10.0\n",
      "suds-community==1.1.2\n",
      "svgwrite==1.4.3\n",
      "tables==3.8.0\n",
      "tabulate==0.9.0\n",
      "tblib==3.0.0\n",
      "tenacity==8.2.3\n",
      "terminado==0.18.0\n",
      "texttable @ file:///home/conda/feedstock_root/build_artifacts/texttable_1696407540773/work\n",
      "threadpoolctl==3.2.0\n",
      "tifffile==2023.7.10\n",
      "tinycss2==1.2.1\n",
      "tmtoolkit==0.12.0\n",
      "tomli==2.0.1\n",
      "toolz==0.12.1\n",
      "tornado @ file:///croot/tornado_1696936946304/work\n",
      "tqdm==4.66.1\n",
      "traitlets @ file:///croot/traitlets_1671143879854/work\n",
      "Tree==0.2.4\n",
      "tspex==0.6.3\n",
      "types-python-dateutil==2.8.19.20240106\n",
      "typing==3.7.4.3\n",
      "typing-inspect==0.9.0\n",
      "typing_extensions @ file:///croot/typing_extensions_1705599297034/work\n",
      "tzdata==2023.4\n",
      "umap-learn==0.5.5\n",
      "uri-template==1.3.0\n",
      "url-normalize==1.4.3\n",
      "urllib3==2.2.0\n",
      "virtualenv==20.21.0\n",
      "wcwidth @ file:///Users/ktietz/demo/mc3/conda-bld/wcwidth_1629357192024/work\n",
      "webcolors==1.13\n",
      "webencodings==0.5.1\n",
      "websocket-client==1.7.0\n",
      "widgetsnbextension==4.0.9\n",
      "wrapt==1.16.0\n",
      "xlrd==2.0.1\n",
      "xmltodict==0.13.0\n",
      "xyzservices==2023.10.1\n",
      "yarl==1.9.4\n",
      "zict==3.0.0\n",
      "zipp @ file:///croot/zipp_1704206909481/work\n",
      "zope.event==5.0\n",
      "zope.interface==6.1\n"
     ]
    }
   ],
   "source": [
    "!pip freeze #Python packages and versions used in the conda environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "29d14293-6fb6-4580-9ab1-d99f7aefdfb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: deepTools\n",
      "Version: 3.5.4.post1\n",
      "Summary: Useful tools for exploring deep sequencing data.\n",
      "Home-page: \n",
      "Author: Fidel Ramirez, Devon P Ryan, Björn Grüning, Friederike Dündar, Sarah Diehl, Vivek Bhardwaj, Fabian Kilpert, Andreas S Richter, Steffen Heyne, Thomas Manke\n",
      "Author-email: bioinfo-core@ie-freiburg.mpg.de\n",
      "License: The file deeptools/cm.py is licensed under the BSD license, see a copy in that file. The remainder of the code is licensed under the MIT license:\n",
      "        \n",
      "        Copyright 2019 Max Planck Institute for Immunobiology and Epigenetics\n",
      "        \n",
      "        Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the \"Software\"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:\n",
      "        \n",
      "        The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.\n",
      "        \n",
      "        THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n",
      "        \n",
      "Location: /root/miniconda3/envs/SCENICplus/lib/python3.8/site-packages\n",
      "Requires: deeptoolsintervals, importlib-metadata, matplotlib, numpy, numpydoc, plotly, py2bit, pyBigWig, pysam, scipy\n",
      "Required-by: \n"
     ]
    }
   ],
   "source": [
    "!pip show deeptools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7e05ee47-7ec5-4219-8bcb-d96a950666ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wigToBigWig v 2.9 - Convert ascii format wig file (in fixedStep, variableStep\n",
      "or bedGraph format) to binary big wig format (bbi version: 4).\n",
      "usage:\n",
      "   wigToBigWig in.wig chrom.sizes out.bw\n",
      "Where in.wig is in one of the ascii wiggle formats, but not including track lines\n",
      "and chrom.sizes is a two-column file/URL: <chromosome name> <size in bases>\n",
      "and out.bw is the output indexed big wig file.\n",
      "If the assembly <db> is hosted by UCSC, chrom.sizes can be a URL like\n",
      "  http://hgdownload.soe.ucsc.edu/goldenPath/<db>/bigZips/<db>.chrom.sizes\n",
      "or you may use the script fetchChromSizes to download the chrom.sizes file.\n",
      "If not hosted by UCSC, a chrom.sizes file can be generated by running\n",
      "twoBitInfo on the assembly .2bit file.\n",
      "options:\n",
      "   -blockSize=N - Number of items to bundle in r-tree.  Default 256\n",
      "   -itemsPerSlot=N - Number of data points bundled at lowest level. Default 1024\n",
      "   -clip - If set just issue warning messages rather than dying if wig\n",
      "                  file contains items off end of chromosome or chromosomes\n",
      "                  that are not in the chrom.sizes file.\n",
      "   -unc - If set, do not use compression.\n",
      "   -fixedSummaries - If set, use a predefined sequence of summary levels.\n",
      "   -keepAllChromosomes - If set, store all chromosomes in b-tree.\n"
     ]
    }
   ],
   "source": [
    "!./wigToBigWig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e2a31bb1-4ad9-4d68-9049-f689f3f1ee00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sort-bed\n",
      "  citation: http://bioinformatics.oxfordjournals.org/content/28/14/1919.abstract\n",
      "            https://doi.org/10.1093/bioinformatics/bts277\n",
      "  version:  2.4.37 (typical)\n",
      "  authors:  Scott Kuehn\n"
     ]
    }
   ],
   "source": [
    "!sort-bed --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d1fed00b-16b3-42ce-890b-37c2527bf672",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sort (GNU coreutils) 8.30\n",
      "Copyright (C) 2018 Free Software Foundation, Inc.\n",
      "License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.\n",
      "This is free software: you are free to change and redistribute it.\n",
      "There is NO WARRANTY, to the extent permitted by law.\n",
      "\n",
      "Written by Mike Haertel and Paul Eggert.\n"
     ]
    }
   ],
   "source": [
    "!sort --version"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deeptools-kernel",
   "language": "python",
   "name": "deeptools-kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
